Copyright 2020 Andrew M. Olney and made available under [CC BY-SA](https://creativecommons.org/licenses/by-sa/4.0) for text and [Apache-2.0](http://www.apache.org/licenses/LICENSE-2.0) for code.

# TEMPLATE THIS NOTEBOOK TO MAKE FAR+ CONDITIONS
  
---------------

## Background

We will be working with data on passengers on the Titanic. The data has eight variables:

- `Survived`: did the passenger survive (0=no, 1=yes)
- `Pclass`: ticket class (1=1st, 2=2nd, 3=3rd)
- `Name`: name of passenger
- `Sex` : sex of passenger
- `Age` : age of passenger
- `Siblings/Spouses Aboard` : number of siblings or spouses aboard
- `Parents/Children Aboard` : number of parents or children aboard
- `Fare` : fare paid in dollars

In this session, you will solve several problems using this data.

<div class="alert alert-warning">
    
**Condition alternative instructions: Goal structure labels**

### Read CSV into a dataframe
   
</div>

Import the `pandas` library, which lets us work with dataframes.

In [12]:
import pandas as pd

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="i!#]:2XI=^qLb$e.|iwo">pd</variable></variables><block type="importAs" id="8u3elQqk_!6!WoHrlj}e" x="73" y="63"><field name="libraryName">pandas</field><field name="libraryAlias" id="i!#]:2XI=^qLb$e.|iwo">pd</field></block></xml>

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>

Load a dataframe with the data in "datasets/titanic.csv" and display it

In [13]:
dataframe = pd.read_csv('datasets/titanic.csv')

dataframe

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="t[n^Fcp7,s93E17ZZ9J6">dataframe</variable><variable id="i!#]:2XI=^qLb$e.|iwo">pd</variable></variables><block type="variables_set" id="YMNSvjU:9aS0`rADBobh" x="29" y="215"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field><value name="VALUE"><block type="varDoMethod" id="h[BIiU^0[[vbD`zoBn6+"><field name="VAR" id="i!#]:2XI=^qLb$e.|iwo">pd</field><field name="MEMBER">read_csv</field><data>pd:read_csv</data><value name="INPUT"><block type="text" id="HyH?(x3/MuPXE`T5;)[@"><field name="TEXT">datasets/titanic.csv</field></block></value></block></value></block><block type="variables_get" id="uS,Sc{|xLGBqCK`F3*-*" x="8" y="300"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field></block></xml>

Survived  Pclass  ... Parents/Children Aboard     Fare
0           0       3  ...                       0   7.2500
1           1       1  ...                       0  71.2833
2           1       3  ...                       0   7.9250
3           1       1  ...                       0  53.1000
4           0       3  ...                       0   8.0500
..        ...     ...  ...                     ...      ...
882         0       2  ...                       0  13.0000
883         1       1  ...                       0  30.0000
884         0       3  ...                       2  23.4500
885         1       1  ...                       0  30.0000
886         0       3  ...                       0   7.7500

[887 rows x 8 columns]

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>

<div class="alert alert-warning">
    
**Condition alternative instructions: Goal structure labels**

### Select rows from a dataframe by position
   
</div>

Show the 4th row to the 885th row (total 882 rows).

In [15]:
dataframe[3 : 885]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="t[n^Fcp7,s93E17ZZ9J6">dataframe</variable></variables><block type="lists_getSublist" id="5hFiqAq|QPxp%Xc:h/@A" x="8" y="518"><mutation at1="true" at2="true"></mutation><field name="WHERE1">FROM_START</field><field name="WHERE2">FROM_START</field><value name="LIST"><block type="variables_get" id="Y`nWHnCXvg4!GpZ![Z8S"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field></block></value><value name="AT1"><block type="math_number" id=":;n{~x;%6oD9(oxp_*Pt"><field name="NUM">4</field></block></value><value name="AT2"><block type="math_number" id="fVh*!`:1z(oQ0WlzO,D("><field name="NUM">885</field></block></value></block></xml>

Survived  Pclass  ... Parents/Children Aboard     Fare
3           1       1  ...                       0  53.1000
4           0       3  ...                       0   8.0500
5           0       3  ...                       0   8.4583
6           0       1  ...                       0  51.8625
7           0       3  ...                       1  21.0750
..        ...     ...  ...                     ...      ...
880         0       3  ...                       0   7.0500
881         0       3  ...                       5  29.1250
882         0       2  ...                       0  13.0000
883         1       1  ...                       0  30.0000
884         0       3  ...                       2  23.4500

[882 rows x 8 columns]

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>

<div class="alert alert-warning">
    
**Condition alternative instructions: Goal structure labels**

### Select columns from a dataframe by name
   
</div>

Show the last column and the first column of the data, in that order.

In [16]:
dataframe[['Fare', 'Survived']]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="t[n^Fcp7,s93E17ZZ9J6">dataframe</variable></variables><block type="indexer" id="oVNd/g7vyxV[^cAbT$JX" x="8" y="300"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field><value name="INDEX"><block type="lists_create_with" id="^oGt7#=i@OMTRwaf:kdd"><mutation items="2"></mutation><value name="ADD0"><block type="text" id="=k__Y{.gxL5z.AJpll.A"><field name="TEXT">Fare</field></block></value><value name="ADD1"><block type="text" id="uQyKHkAg(%zJh!AAX=[#"><field name="TEXT">Survived</field></block></value></block></value></block></xml>

Fare  Survived
0     7.2500         0
1    71.2833         1
2     7.9250         1
3    53.1000         1
4     8.0500         0
..       ...       ...
882  13.0000         0
883  30.0000         1
884  23.4500         0
885  30.0000         1
886   7.7500         0

[887 rows x 2 columns]

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>

<div class="alert alert-warning">
    
**Condition alternative instructions: Goal structure labels**

### Select rows from a dataframe by value
    
</div>

Show the rows where `Sex` is female.

In [17]:
dataframe[(dataframe['Sex'] == 'female')]

#<xml xmlns="https://developers.google.com/blockly/xml"><variables><variable id="t[n^Fcp7,s93E17ZZ9J6">dataframe</variable></variables><block type="indexer" id="Y*eD.jv!V*/5.i:+x([t" x="19" y="390"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field><value name="INDEX"><block type="logic_compare" id="FbNIQ=Q,*rY%1.XvHA^q"><field name="OP">EQ</field><value name="A"><block type="indexer" id="d3u5qTTbrdfW-F2:-QR2"><field name="VAR" id="t[n^Fcp7,s93E17ZZ9J6">dataframe</field><value name="INDEX"><block type="text" id="rX][Xdf,hW{pZ4}3u$Sq"><field name="TEXT">Sex</field></block></value></block></value><value name="B"><block type="text" id="D%gs;.Z]qVUL%%ev5qEE"><field name="TEXT">female</field></block></value></block></value></block></xml>

Survived  Pclass  ... Parents/Children Aboard     Fare
1           1       1  ...                       0  71.2833
2           1       3  ...                       0   7.9250
3           1       1  ...                       0  53.1000
8           1       3  ...                       2  11.1333
9           1       2  ...                       0  30.0708
..        ...     ...  ...                     ...      ...
876         1       2  ...                       1  26.0000
878         0       3  ...                       0  10.5167
881         0       3  ...                       5  29.1250
883         1       1  ...                       0  30.0000
884         0       3  ...                       2  23.4500

[314 rows x 8 columns]

<div class="alert alert-danger">
    
**Condition alternative instructions: Self-explanation**

Self-explain the code/output for this cell: what is it and what does it do?
    
</div>